### PR TITLE
gemma4: batch multimodal vision processing

### PR DESCRIPTION
**Pull request by wuyingjun-lucky**
## Summary
- batch Gemma 4 image vision-tower processing and split the flat output back into per-item embeddings using pooled token counts
- add focused processing tests for pooled token count reconstruction
- guard SM100-only MXFP4 stable libtorch bindings behind `ENABLE_NVFP4_SM100`
- chunk Gemma 4 video vision profiling batches to avoid regressing 16GB E2B startup memory on smaller GPUs

## Testing
- `LD_LIBRARY_PATH=/home/wuyingjun/llm/vllm/.venv/lib/python3.12/site-packages/torch/lib:/usr/local/cuda-13.2/targets/x86_64-linux/lib:/home/wuyingjun/llm/vllm/.venv/lib/python3.12/site-packages/nvidia/cu13/lib:/usr/local/cuda-12.9/lib64 /home/wuyingjun/llm/vllm/.venv/bin/python -m pytest /home/wuyingjun/llm/vllm/tests/models/multimodal/processing/test_gemma4_mm.py`
- rebuilt `_C_stable_libtorch` on local `sm89 + CUDA 12.9` to verify the SM100 guard compiles cleanly
- verified local `google/gemma-4-E2B-it` startup on a 16GB RTX 4070 Ti SUPER with `max_model_len=128` and `gpu_memory_utilization=0.85` after the video chunking fix

### CODE REVIEW

File: vllm/model_executor/models/gemma4_mm.py [hunk 1]
```diff
@@ -1066,65 +1069,70 @@ def _process_image_input(
...
+        if (max_items_per_batch is not None
+                and pixel_values.shape[0] > max_items_per_batch):
+            embeddings: list[torch.Tensor] = []
+            for start in range(0, pixel_values.shape[0], max_items_per_batch):
+                end = start + max_items_per_batch
+                embeddings.extend(
+                    self._process_vision_batch(
+                        pixel_values[start:end],
+                        pixel_position_ids[start:end],
+                    ))
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:1279]: Can you provide some benchmarks results about the performance gains? You can use `vllm bench mm-processor`

Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:1279]: Ran `vllm bench mm-processor` on RTX 4070 Ti SUPER 16GB / Ryzen 7 9700X / WSL2 with 6 prompts, 1 warmup, 1 video per request, bucket `(224, 224, 8)`, `max_model_len=2048`, `--enforce-eager`, and `--gpu-memory-utilization 0.85`.

Baseline (`video_vision_batch_size=1`): `apply_hf_processor_ms=17.13`, `preprocessor_total_ms=17.54`, `encoder_forward_ms=105.61`, `mean_e2el_ms=2002.89`.

Configured (`--mm-processor-kwargs {"video_vision_batch_size": 4}`): `apply_hf_processor_ms=15.42`, `preprocessor_total_ms=15.97`, `encoder_forward_ms=61.15`, `mean_e2el_ms=670.93`.

For this setup that is about `1.73x` faster on `encoder_forward_ms` and about `2.99x` lower mean end-to-end latency. I also fixed a benchmark-only HF warning on this path so `video_vision_batch_size` now stays local to vLLM and no longer leaks into the HF processor call.

**Comment by wuyingjun-lucky** [comment]
Adding the benchmark summary here as a standalone note.

## Benchmark Results (`vllm bench mm-processor`)

Hardware: RTX 4070 Ti SUPER 16GB / Ryzen 7 9700X / WSL2
Workload: 6 prompts, 1 warmup, 1 video per request, bucket `(224, 224, 8)`, `max_model_len=2048`, `--enforce-eager`, `--gpu-memory-utilization 0.85`

| Config | apply_hf_processor (ms) | preprocessor_total (ms) | encoder_forward (ms) | mean_e2e (ms) | Relative change |
|---|---:|---:|---:|---:|---|
| `video_vision_batch_size=1` | 17.13 | 17.54 | 105.61 | 2002.89 | baseline |
| `video_vision_batch_size=4` | 15.42 | 15.97 | 61.15 | 670.93 | `encoder_forward` `1.73x` faster, mean E2E latency `2.99x` lower |

The latest patch also fixes the benchmark-only HF warning so `video_vision_batch_size` now stays local to vLLM and no longer leaks into the HF processor call path.

Please take another look.

File: vllm/model_executor/models/gemma4_mm.py [hunk 2]
```diff
@@ -184,9 +213,19 @@ def get_default_tok_params(self):
...
-        return self.ctx.get_hf_processor(
-            Gemma4Processor,
-            **kwargs,
+        tokenizer = self.ctx.get_tokenizer()
+        if is_mistral_tokenizer(tokenizer):
+            tokenizer = tokenizer.transformers_tokenizer
+
+        merged_kwargs = self.ctx.get_merged_mm_kwargs(kwargs)
+        merged_kwargs = _get_hf_processor_mm_kwargs(merged_kwargs)
+        merged_kwargs.pop("tokenizer", None)
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:321]: Is this relative?

File: vllm/model_executor/models/gemma4_mm.py [hunk 3]
```diff
@@ -535,16 +631,15 @@ def _call_hf_processor(
...
+                video_mm_kwargs = dict(hf_mm_kwargs)
                 video_mm_kwargs["max_soft_tokens"] = _VIDEO_MAX_SOFT_TOKENS
 
                 dummy_prompt = ("\t" + processor.image_token) * len(frames)
 
-                frame_outputs = super()._call_hf_processor(
-                    prompt=dummy_prompt,
-                    mm_data={"images": frames},
-                    mm_kwargs=video_mm_kwargs,
-                    tok_kwargs=tok_kwargs,
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:547]: Why can't we use `super()._call_hf_processor()`? I don't think we should add `_call_hf_processor_without_local_kwargs` that's highly duplicate with `super()._call_hf_processor`

File: vllm/model_executor/models/gemma4_mm.py [hunk 4]
```diff
@@ -987,17 +1100,19 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
         pixel_position_ids = kwargs.pop("pixel_position_ids", None)
         image_embeds = kwargs.pop("image_embeds", None)
+        image_num_soft_tokens = kwargs.pop("image_num_soft_tokens", None)
         assert image_embeds is None, "Gemma4 does not support image_embeds."
         if pixel_values is None:
             return None
-        return Gemma4ImagePixelInputs(
-            pixel_values=pixel_values,
-            pixel_position_ids=pixel_position_ids,
-        )
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:1000]: We should use `Gemma4ImagePixelInputs` for tensor schema validation.

File: vllm/model_executor/models/gemma4_mm.py [hunk 5]
```diff
@@ -1015,21 +1130,25 @@ def _parse_and_validate_audio_input(
...
             "pixel_values_videos": pixel_values_videos,
             "pixel_position_ids_videos": pixel_position_ids_videos,
             "video_frame_counts": video_frame_counts,
+            "video_num_soft_tokens": video_num_soft_tokens,
+            "video_vision_batch_size": video_vision_batch_size,
         }
 
     def _parse_and_validate_multimodal_inputs(
         self, **kwargs: object
-    ) -> dict[str, Gemma4ImageInputs | Gemma4AudioInputs | Gemma4VideoInputs | None]:
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:1032]: Please revert this.

File: vllm/model_executor/models/gemma4_mm.py [hunk 6]
```diff
@@ -1061,78 +1180,89 @@ def _parse_and_validate_multimodal_inputs(
...
-            max_patches = pv.shape[1]
-            output_length = max_patches // pooling_k2
+    def _process_vision_batch(
+        self,
+        pixel_values: torch.Tensor,
+        pixel_position_ids: torch.Tensor,
+        max_items_per_batch: int | None = None,
+        token_counts: list[int] | None = None,
+    ) -> list[torch.Tensor]:
+        """Run the Gemma4 vision tower on a batch and split flat outputs.
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm.py:1252]: Since Gemma-4's ViT is quite small, I think feeding all videos in one batch should be fine?

File: vllm/model_executor/models/gemma4_mm_utils.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+import math
+
+import torch
+
+
+def get_gemma4_pooled_token_counts(
+    pixel_position_ids: torch.Tensor,
+    output_length: int,
+) -> torch.Tensor:
```
Comment (Isotr0py) [inline vllm/model_executor/models/gemma4_mm_utils.py]: We can move this to `gemma4_mm.py`

File: vllm/model_executor/models/gemma4_mm.py (hunk 2 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:321]: Yes. This now mirrors `InputProcessingContext.get_hf_processor()`, but strips vLLM-local kwargs before processor construction so `video_vision_batch_size` never reaches the HF processor.

File: vllm/model_executor/models/gemma4_mm.py (hunk 3 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:547]: Switched back to super()._call_hf_processor() and removed the duplicate local helper. The wrapper only keeps video_vision_batch_size visible to the vLLM kwarg filter, then drops it before the actual HF call.

File: vllm/model_executor/models/gemma4_mm.py (hunk 4 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:1000]: Restored Gemma4ImagePixelInputs for validation and kept image_num_soft_tokens as an optional field on that schema.

File: vllm/model_executor/models/gemma4_mm.py (hunk 5 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:1032]: Reverted this. Video parsing now returns Gemma4VideoInputs again and keeps the extra video fields on the tensor schema.

File: vllm/model_executor/models/gemma4_mm_utils.py (hunk 7 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm_utils.py]: Moved get_gemma4_pooled_token_counts back into gemma4_mm.py and updated the test import accordingly.

File: vllm/model_executor/models/gemma4_mm.py (hunk 6 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:1252]: I kept the default at 1 for small-GPU startup safety because the engine profiles max-size video items during init. Runtime callers can still opt into larger batches via mm_processor_kwargs. I reran vllm bench mm-processor on the (224, 224, 8) video workload after this change: baseline encoder_forward_ms=123.90 and mean_e2el_ms=4753.17; video_vision_batch_size=4 gives encoder_forward_ms=60.61 and mean_e2el_ms=2852.99, with no HF warning on this path.

File: vllm/model_executor/models/gemma4_mm.py (hunk 1 above)
Comment (wuyingjun-lucky) [reply vllm/model_executor/models/gemma4_mm.py:1279]: Reran vllm bench mm-processor after the latest review-fix patch on the same 6 prompt / 1 warmup / (224, 224, 8) video workload with max_model_len=2048, enforce_eager, and gpu_memory_utilization=0.85. Baseline: apply_hf_processor_ms=18.73, preprocessor_total_ms=20.37, encoder_forward_ms=123.90, mean_e2el_ms=4753.17. Configured video_vision_batch_size=4: apply_hf_processor_ms=12.44, preprocessor_total_ms=13.81, encoder_forward_ms=60.61, mean_e2el_ms=2852.99. That is about 2.04x faster on encoder_forward_ms and about 1.67x lower mean end-to-end latency, and the local-only kwarg no longer leaks into the HF processor call.
